src/vision/status_quo/grace_wants_to_integrate_c_api.md

```diff
@@ -0,0 +1,176 @@
+of some kind, so that her team can write something similar to:
+
+```rust
+let frame_stream = stream_from_camera(camera, Duration::from_millis(5));
+
+while let Some(frame) = frame_stream.next().await {
+    // process frames
+}
+
+println!("Frame stream closed.");
+```
+
+She scours existing crates, in particular looking for one way to transform the above future into a stream that can be
+executed many times. The only available option to transform a future into a series of futures is [`stream::unfold`],
+which seems to do exactly what Grace is looking for. Grace begins by adding a small intermediate type, and then plugging
+in the remaining holes:
+
+```rust
+struct StreamState {
+    camera: Camera,
+    retry_after: Duration,
+}
+
+
+
+fn stream_from_camera(camera: Camera, retry_after: Duration) -> Unfold<Frame, ??, ??> {
+    let initial_state = StreamState { camera, retry_after };
+
+    stream::unfold(initial_state, |state| async move {
+        let frame = next_frame(&state.camera, state.retry_after).await
+        (frame, state)
+    })
+}
+```
+
+This looks like it mostly hits the mark, but Grace is left with a couple of questions for how to get the remainder of
+this building:
+
+1. What is the type that fills in the third template parameter in the return? It should be the type of the future that
+   is returned by the async closure passed into `stream::unfold`, but we don't know the type of a closure!
+2. What is the type that fills in the second template parameter of the closure in the return?
+
+Grace spends a lot of time trying to figure out how she might find those types! She asks [Barbara] what the idiomatic
+way to get around this in Rust would be. Barbara explains again how closures don't have concrete types, and that the
+only way to do this will be to use the `impl` keyword.
+
+```rust
+fn stream_from_camera(camera: Camera, retry_after: Duration) -> impl Stream<Item = Frame> {
+    // same as before
+}
+```
+
+While Grace was was on the correct path and now her team is able to write the code they want to, she realizes that
+sometimes writing the types out explicitly can be very hard. She reflects on what it would have taken to write the type
+of an equivalent function pointer in C, and slightly laments that Rust cannot express such as clearly.
+
+## 🤔 Frequently Asked Questions
+
+* **What are the morals of the story?**
+    * Rust was the correct choice for the team across the board thanks to its memory safety and ownership. The
+      underlying C library was just too complex for any single programmer to be able to maintain in their head all at
+      once while also trying to accomplish other tasks.
+    * Evolving requirements meant that the team would have had to either start over in plain C, giving up a lot of the
+      safety they would gain from switching to Rust, or exploring async code in a more rigorous way.
```

Comment:
I'm not sure I see this moral reflected in the story. Can you talk a bit more about what you mean by this and how it relates to the story?

Reply:
Thanks for the feedback! I probably did not expound on this too much in the story. 

In reality, when you try to switch towards a new(er) language, there's always the "oops we hit the first roadblock, maybe this whole idea was a mistake." This sentiment happens a lot, especially when switching away from a technology as tried and tested as C. 

I suppose at the point in the story where requirements evolved (1 camera -> 5 cameras), I've seen hesitation to walk back other ambitious parts of a project in order to conserve time or capture risk. This is a bit nuanced, and probably isn't the same across every org! I fully admit that this "moral" is perhaps ill-placed or not-well-explained in this context. In my own experience, we have had debates around whether it was even worth diving into a bigger solution with async given that the project has shifted. Ultimately, I pushed for continuing in Rust because the benefits applied to memory safety could _also_ be true for async, if we could figure out what our unknown unknowns were.

With that in mind, does this moral detract from the overall story? Is the connection too soft? I'm happy to remove it, I think the story as it stands has a lot already, but I could also see adding a small amount of content to help clarify.

Reply:
I think it would be helpful to just add text around this. Something like:

The requirements shifted. Some folks argued for moving away from Rust -- maybe they should stick to the tried and true. But ultimately they decided that the benefits of memory safety were still applicable xxx

Reply:
👍🏻 I added a note in text for this in  c5891dee9f2fa9230c2799d66bf75887fc6b8a02. Let me know if it needs any more fine-tuning!

Reply:
Looks good to me!